internal/db/merge.go

```diff
@@ -0,0 +1,418 @@
+type mergeProcessor struct {
+	txn        datastore.Txn
+	ls         linking.LinkSystem
+	mCRDTs     map[string]merklecrdt.MerkleCRDT
+	col        *collection
+	dsKey      core.DataStoreKey
+	composites *list.List
+}
+
+func (db *db) newMergeProcessor(
+	txn datastore.Txn,
+	ls linking.LinkSystem,
+	col *collection,
+	dsKey core.DataStoreKey,
+) (*mergeProcessor, error) {
+	return &mergeProcessor{
+		txn:        txn,
+		ls:         ls,
+		mCRDTs:     make(map[string]merklecrdt.MerkleCRDT),
+		col:        col,
+		dsKey:      dsKey,
+		composites: list.New(),
+	}, nil
+}
+
+type mergeTarget struct {
+	heads      map[cid.Cid]*coreblock.Block
+	headHeigth uint64
+}
+
+func newMergeTarget() mergeTarget {
+	return mergeTarget{
+		heads: make(map[cid.Cid]*coreblock.Block),
+	}
+}
+
+// loadComposites retrieves and stores into the merge processor the composite blocks for the given
+// document until it reaches a block that has already been merged or until we reach the genesis block.
+func (mp *mergeProcessor) loadComposites(
+	ctx context.Context,
+	blockCid cid.Cid,
+	mt mergeTarget,
+) error {
+	nd, err := mp.ls.Load(linking.LinkContext{Ctx: ctx}, cidlink.Link{Cid: blockCid}, coreblock.SchemaPrototype)
+	if err != nil {
+		return err
+	}
+
+	block, err := coreblock.GetFromNode(nd)
+	if err != nil {
+		return err
+	}
+
+	if _, ok := mt.heads[blockCid]; ok {
+		// We've already processed this block.
+		return nil
+	}
+
+	if block.Delta.GetPriority() >= mt.headHeigth {
+		mp.composites.PushFront(block)
+		for _, link := range block.Links {
+			if link.Name == core.HEAD {
+				err := mp.loadComposites(ctx, link.Cid, mt)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	} else {
+		newMT := newMergeTarget()
+		for _, b := range mt.heads {
+			for _, link := range b.Links {
+				nd, err := mp.ls.Load(linking.LinkContext{Ctx: ctx}, link.Link, coreblock.SchemaPrototype)
+				if err != nil {
+					return err
+				}
+
+				childBlock, err := coreblock.GetFromNode(nd)
+				if err != nil {
+					return err
+				}
+
+				newMT.heads[link.Cid] = childBlock
+				newMT.headHeigth = childBlock.Delta.GetPriority()
+			}
+		}
+		return mp.loadComposites(ctx, blockCid, newMT)
+	}
+	return nil
+}
+
+func (mp *mergeProcessor) mergeComposites(ctx context.Context) error {
+	for e := mp.composites.Front(); e != nil; e = e.Next() {
+		block := e.Value.(*coreblock.Block)
+		link, err := block.GenerateLink()
+		if err != nil {
+			return err
+		}
+		err = mp.processBlock(ctx, block, link)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// processBlock merges the block and its children to the datastore and sets the head accordingly.
+func (mp *mergeProcessor) processBlock(
+	ctx context.Context,
+	block *coreblock.Block,
+	blockLink cidlink.Link,
+) error {
+	crdt, err := mp.initCRDTForType(block.Delta.GetFieldName())
+	if err != nil {
+		return err
+	}
+
+	// If the CRDT is nil, it means the field is not part
+	// of the schema and we can safely ignore it.
```

Comment:
question: When can this happen?  The only time I can think of is if the block was created on a higher schema version with a new field, and then we would still want the block synced - I cant remember if this was fully working, or if there is a ticket for handling the datastore, but I thought the blocks would still be synced (they affect the composite cid, so I don't think it is right to drop them even if we cant yet handle the datastore side).

Reply:
This is not the block sync process. That has already been taken care of in the `net` package. Here we merge the synced blocks into the datastore. Like you said, this can only happen if the block was created on a different schema version. If the node merge doesn't know about this field, it can't merge it into the datastore.